File: desktop/LegionControl.Desktop/Model/SetupMerge.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using LegionControl.Desktop.Config;

namespace LegionControl.Desktop.Model;

/// One thing two copies of the setup disagree about.
///
/// The granularity is the thing a person recognises: a machine, a site, one address on a machine,
/// its wake block, one of its systems, the update repository. A diff at the level of JSON paths
/// would be accurate and unreadable, and "the setups differ" would be readable and useless.
public sealed record SetupDifference(
    string Key,
    string Description,
    string? Mine,
    string? Theirs,
    SetupChange Change)
{
    /// Whether both sides moved this entry away from their common ancestor. Those are the only
    /// ones a person has to answer for; everything else merges without asking.
    public bool NeedsChoice => Change == SetupChange.Both;
}

public enum SetupChange
{
    /// Only this device changed it, so the merge keeps this side.
    Mine,
    /// Only the other side changed it, so the merge takes theirs.
    Theirs,
    /// Both changed it, or there is no common ancestor to tell which did.
    Both,
}

/// Which side of a difference to keep.
public enum SetupChoice
{
    Mine,
    Theirs,
}

/// Compares and merges two setup documents.
///
/// This is what makes concurrent offline edits survive. Two peers that both edited revision 5 have
/// two revision 6s, and neither number nor hash can say which to keep: the answer is to keep both
/// people's work, entry by entry, and to ask about the entries where that is impossible.
public static class SetupMerge
{
    /// The entries two documents disagree about, judged against their common ancestor when there is
    /// one. Without a base every difference is [SetupChange.Both], which is honest: nobody can say
    /// who moved what.
    public static IReadOnlyList<SetupDifference> Compare(
        ControllerDocument mine, ControllerDocument theirs, ControllerDocument? baseDocument)
    {
        var mineEntries = Entries(mine);
        var theirEntries = Entries(theirs);
        var baseEntries = baseDocument is null ? null : Entries(baseDocument);

        var keys = mineEntries.Keys.Union(theirEntries.Keys, StringComparer.Ordinal)
            .OrderBy(key => key, StringComparer.Ordinal);
        var differences = new List<SetupDifference>();

        foreach (var key in keys)
        {
            mineEntries.TryGetValue(key, out var mineValue);
            theirEntries.TryGetValue(key, out var theirValue);
            if (mineValue == theirValue) continue;

            var change = SetupChange.Both;
            if (baseEntries is not null)
            {
                baseEntries.TryGetValue(key, out var baseValue);
                var mineMoved = mineValue != baseValue;
                var theirsMoved = theirValue != baseValue;
                change = (mineMoved, theirsMoved) switch
                {
                    (true, false) => SetupChange.Mine,
                    (false, true) => SetupChange.Theirs,
                    _ => SetupChange.Both,
                };
            }

            differences.Add(new SetupDifference(key, Describe(key, mineValue, theirValue), mineValue, theirValue, change));
        }
        return differences;
    }

    /// Builds the merged document.
    ///
    /// Entries only one side touched are taken from that side. Entries both sides touched follow
    /// [choices], which the divergence sheet fills in and which defaults to theirs, because the
    /// side that is already on the machines is the one a silent default should not overwrite.
    public static ControllerDocument Merge(
        ControllerDocument mine, string mineHash,
        ControllerDocument theirs, string theirHash,
        ControllerDocument? baseDocument,
        IReadOnlyDictionary<string, SetupChoice>? choices,
        string? deviceName)
    {
        var differences = Compare(mine, theirs, baseDocument);
        var result = JsonNode.Parse(mine.Text) as JsonObject ?? new JsonObject();
        var theirRoot = JsonNode.Parse(theirs.Text) as JsonObject ?? new JsonObject();

        foreach (var difference in differences)
        {
            var take = difference.Change switch
            {
                SetupChange.Mine => SetupChoice.Mine,
                SetupChange.Theirs => SetupChoice.Theirs,
                _ => choices is not null && choices.TryGetValue(difference.Key, out var chosen)
                    ? chosen
                    : SetupChoice.Theirs,
            };
            if (take == SetupChoice.Mine) continue;
            ApplyFrom(result, theirRoot, difference.Key);
        }

        var identity = ControllerIdentity.Merged(mine.Identity, mineHash, theirs.Identity, theirHash, deviceName);
        result["controller"] = ConfigStore.IdentityJson(identity, result["controller"] as JsonObject);
        return ControllerDocument.FromText(result.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
    }

    /// The comparable entries of a document, keyed by the path a person would name.
    ///
    /// The `controller` block is deliberately absent: it is bookkeeping about the document rather
    /// than part of the setup, and every merge writes a fresh one.
    internal static Dictionary<string, string> Entries(ControllerDocument document)
    {
        var entries = new Dictionary<string, string>(StringComparer.Ordinal);
        if (JsonNode.Parse(document.Text) is not JsonObject root) return entries;

        foreach (var property in root)
        {
            switch (property.Key)
            {
                case "controller":
                    continue;
                case "machines":
                    foreach (var entry in property.Value as JsonArray ?? new JsonArray())
                    {
                        if (entry is not JsonObject machine) continue;
                        var id = machine["id"]?.GetValue<string>() ?? "?";
                        foreach (var part in machine)
                        {
                            if (part.Key == "id") continue;
                            if (part.Key == "endpoints")
                            {
                                foreach (var address in part.Value as JsonArray ?? new JsonArray())
                                {
                                    if (address is not JsonObject endpoint) continue;
                                    var addressId = endpoint["id"]?.GetValue<string>()
                                                    ?? endpoint["host"]?.GetValue<string>() ?? "?";
                                    entries[$"machines[{id}].endpoints[{addressId}]"] = Canonicalise(endpoint);
                                }
                                continue;
                            }
                            if (part.Key == "systems")
                            {
                                foreach (var entrySystem in part.Value as JsonArray ?? new JsonArray())
                                {
                                    if (entrySystem is not JsonObject system) continue;
                                    var systemId = system["id"]?.GetValue<string>() ?? "?";
                                    entries[$"machines[{id}].systems[{systemId}]"] = Canonicalise(system);
                                }
                                continue;
                            }
                            entries[$"machines[{id}].{part.Key}"] = Canonicalise(part.Value);
                        }
                    }
                    continue;
                case "sites":
                    foreach (var entry in property.Value as JsonArray ?? new JsonArray())
                    {
                        if (entry is not JsonObject site) continue;
                        entries[$"sites[{site["id"]?.GetValue<string>() ?? "?"}]"] = Canonicalise(site);
                    }
                    continue;
                default:
                    entries[property.Key] = Canonicalise(property.Value);
                    continue;
            }
        }
        return entries;
    }

    /// Copies one entry from one document into another, creating the machine, address or system it
    /// belongs to when this side does not have it, and removing it when the other side does not.
    private static void ApplyFrom(JsonObject target, JsonObject source, string key)
    {
        if (!key.StartsWith("machines[", StringComparison.Ordinal) && !key.StartsWith("sites[", StringComparison.Ordinal))
        {
            var value = source[key];
            if (value is null) target.Remove(key);
            else target[key] = value.DeepClone();
            return;
        }

        if (key.StartsWith("sites[", StringComparison.Ordinal))
        {
            var id = Between(key, "sites[", "]");
            ReplaceInArray(target, source, "sites", id, null);
            return;
        }

        var machineId = Between(key, "machines[", "]");
        var rest = key[(key.IndexOf(']') + 2)..];

        var targetMachine = FindOrAdd(target, "machines", machineId);
        var sourceMachine = Find(source, "machines", machineId);
        if (sourceMachine is null)
        {
            // The other side does not have this machine at all, so taking their side of a
            // difference about it means removing it here.
            RemoveFromArray(target, "machines", machineId);
            return;
        }

        if (rest.StartsWith("endpoints[", StringComparison.Ordinal))
        {
            ReplaceInArray(targetMachine, sourceMachine, "endpoints", Between(rest, "endpoints[", "]"), "host");
            return;
        }
        if (rest.StartsWith("systems[", StringComparison.Ordinal))
        {
            ReplaceInArray(targetMachine, sourceMachine, "systems", Between(rest, "systems[", "]"), null);
            return;
        }
        var part = sourceMachine[rest];
        if (part is null) targetMachine.Remove(rest);
        else targetMachine[rest] = part.DeepClone();
    }

    private static void ReplaceInArray(JsonObject target, JsonObject source, string arrayName, string id, string? fallbackKey)
    {
        var sourceArray = source[arrayName] as JsonArray ?? new JsonArray();
        var replacement = sourceArray.FirstOrDefault(entry =>
            entry is JsonObject item
            && (item["id"]?.GetValue<string>() ?? (fallbackKey is null ? null : item[fallbackKey]?.GetValue<string>())) == id);

        if (target[arrayName] is not JsonArray targetArray)
        {
            if (replacement is null) return;
            targetArray = new JsonArray();
            target[arrayName] = targetArray;
        }

        for (var index = targetArray.Count - 1; index >= 0; index -= 1)
        {
            if (targetArray[index] is not JsonObject item) continue;
            var itemId = item["id"]?.GetValue<string>()
                         ?? (fallbackKey is null ? null : item[fallbackKey]?.GetValue<string>());
            if (itemId != id) continue;
            if (replacement is null) targetArray.RemoveAt(index);
            else targetArray[index] = replacement.DeepClone();
            return;
        }
        if (replacement is not null) targetArray.Add(replacement.DeepClone());
    }

    private static JsonObject FindOrAdd(JsonObject root, string arrayName, string id)
    {
        if (Find(root, arrayName, id) is { } existing) return existing;
        if (root[arrayName] is not JsonArray array)
        {
            array = new JsonArray();
            root[arrayName] = array;
        }
        var created = new JsonObject { ["id"] = id };
        array.Add(created);
        return created;
    }

    private static JsonObject? Find(JsonObject root, string arrayName, string id) =>
        (root[arrayName] as JsonArray)?
        .OfType<JsonObject>()
        .FirstOrDefault(entry => entry["id"]?.GetValue<string>() == id);

    private static void RemoveFromArray(JsonObject root, string arrayName, string id)
    {
        if (root[arrayName] is not JsonArray array) return;
        for (var index = array.Count - 1; index >= 0; index -= 1)
        {
            if (array[index] is JsonObject entry && entry["id"]?.GetValue<string>() == id) array.RemoveAt(index);
        }
    }

    private static string Between(string text, string prefix, string suffix)
    {
        var start = text.IndexOf(prefix, StringComparison.Ordinal) + prefix.Length;
        var end = text.IndexOf(suffix, start, StringComparison.Ordinal);
        return end < 0 ? text[start..] : text[start..end];
    }

    /// A stable string for one value, so two documents that say the same thing in a different key
    /// order compare equal.
    private static string Canonicalise(JsonNode? node)
    {
        if (node is null) return "";
        if (node is JsonObject entry)
        {
            var ordered = new JsonObject();
            foreach (var property in entry.OrderBy(property => property.Key, StringComparer.Ordinal))
            {
                ordered[property.Key] = property.Value is null ? null : JsonNode.Parse(Canonicalise(property.Value));
            }
            return ordered.ToJsonString();
        }
        if (node is JsonArray array)
        {
            var copy = new JsonArray();
            foreach (var item in array) copy.Add(item is null ? null : JsonNode.Parse(Canonicalise(item)));
            return copy.ToJsonString();
        }
        return node.ToJsonString();
    }

    private static string Describe(string key, string? mine, string? theirs)
    {
        if (mine is null) return $"{key} exists only on the other side.";
        if (theirs is null) return $"{key} exists only here.";
        return $"{key} differs.";
    }
}

File: mac/Sources/LegionControlCore/SetupMerge.swift
import Foundation

/// Putting two diverged setups back together.
///
/// When both sides changed the document since they last agreed, neither may simply overwrite the
/// other. What this does is compare them entry by entry — one machine, one site, one endpoint, one
/// system, the app update settings — and, where a common ancestor is available, work out which side
/// actually changed each entry. Entries only one side touched are taken from that side without
/// asking. Entries both sides touched are the only ones a person has to decide about, and they are
/// shown with both versions.
///
/// Everything works on the raw JSON rather than on the typed model, and that is deliberate: a
/// document may legitimately carry keys this build has never heard of — a newer client wrote them,
/// or only the phone reads them — and round-tripping through `ControllerConfig` would silently
/// delete every one of them.
enum SetupMerge {

    /// One thing the two documents disagree about.
    struct Difference: Sendable, Equatable, Identifiable {
        /// Where it is, as a path a person can read: `machines.legion.wake`, `sites.home-a`.
        var path: [String]
        var kind: Kind
        /// What is on this device, as pretty JSON, or nil when this side does not have it.
        var mine: String?
        /// What the other side has, or nil when it does not have it.
        var theirs: String?
        /// What the common ancestor had, when there is one.
        var base: String?

        var id: String { path.joined(separator: ".") }
        var label: String { path.joined(separator: " › ") }

        enum Kind: Sendable, Equatable {
            /// Only this device changed it. Taken from here, no question asked.
            case onlyMineChanged
            /// Only the other side changed it. Taken from there, no question asked.
            case onlyTheirsChanged
            /// Both changed it, differently. The one case a person has to settle.
            case bothChanged
            /// No common ancestor was available, so which side changed it cannot be worked out.
            case unknownBase
        }

        /// What a merge does with it unless the user says otherwise.
        var defaultChoice: Choice {
            switch kind {
            case .onlyMineChanged: .mine
            case .onlyTheirsChanged: .theirs
            // With no evidence about who moved, the copy the machines are already carrying wins.
            // Taking "mine" by default would quietly undo an edit made on another device.
            case .bothChanged, .unknownBase: .theirs
            }
        }

        /// Whether it needs a person. Everything else is settled by the evidence.
        var needsDecision: Bool { kind == .bothChanged || kind == .unknownBase }
    }

    enum Choice: String, Sendable, Equatable, CaseIterable {
        case mine
        case theirs
    }

    struct MergeFailure: Error, Sendable, Equatable {
        var message: String
        init(_ message: String) { self.message = message }
    }

    /// Compare two documents, with the common ancestor when one is to hand.
    static func differences(mine: Data, theirs: Data, base: Data?) throws -> [Difference] {
        let a = try object(mine, what: "this device's setup")
        let b = try object(theirs, what: "the machine's setup")
        let c = base.flatMap { try? object($0, what: "the common ancestor") }

        var out: [Difference] = []
        for path in entryPaths(a: a, b: b, base: c) {
            let mineValue = value(at: path, in: a)
            let theirsValue = value(at: path, in: b)
            if equal(mineValue, theirsValue) { continue }
            let baseValue = c.flatMap { value(at: path, in: $0) }

            let kind: Difference.Kind
            if c == nil {
                kind = .unknownBase
            } else if equal(baseValue, mineValue) {
                kind = .onlyTheirsChanged
            } else if equal(baseValue, theirsValue) {
                kind = .onlyMineChanged
            } else {
                kind = .bothChanged
            }

            out.append(Difference(
                path: path,
                kind: kind,
                mine: mineValue.map(pretty),
                theirs: theirsValue.map(pretty),
                base: baseValue.map(pretty)
            ))
        }
        return out.sorted { $0.id < $1.id }
    }

    /// Build the merged document.
    ///
    /// `choices` is keyed by difference id and only has to cover the ones that need a decision;
    /// anything missing takes the default, which is the side that changed it.
    static func merge(
        mine: Data,
        theirs: Data,
        base: Data?,
        differences: [Difference],
        choices: [String: Choice],
        identity: ControllerIdentity
    ) throws -> Data {
        var result = try object(mine, what: "this device's setup")
        let other = try object(theirs, what: "the machine's setup")

        for difference in differences {
            let choice = choices[difference.id] ?? difference.defaultChoice
            guard choice == .theirs else { continue }
            let theirsValue = value(at: difference.path, in: other)
            result = setting(difference.path, to: theirsValue, in: result)
        }

        // The identity is always this device's own: it descends from both sides, so both machines
        // accept it as a fast-forward and neither branch is lost.
        result["controller"] = encode(identity)

        // Written the way every device writes a document, so the merge has the same bytes and the
        // same hash wherever it was made.
        return try JSONText.canonicalDocument(result)
    }

    /// The identity block as a plain dictionary, so unknown keys elsewhere in the document survive.
    static func encode(_ identity: ControllerIdentity) -> [String: Any] {
        var block: [String: Any] = [:]
        if let id = identity.id { block["id"] = id }
        if let name = identity.name { block["name"] = name }
        block["revision"] = identity.revisionNumber
        if let updatedAt = identity.updatedAt { block["updatedAt"] = updatedAt }
        if let source = identity.source { block["source"] = source }
        if let device = identity.device { block["device"] = device }
        block["lineage"] = identity.ancestors
        return block
    }

    // MARK: - Walking the document

    /// The paths a merge works at.
    ///
    /// Not every key and not the whole document: one machine, one site, one endpoint, one system,
    /// the wake block, the app update settings. Fine enough that renaming a machine on one device
    /// and adding an endpoint to another one elsewhere merges without a question, coarse enough that
    /// a person is never asked about a port number in isolation.
    static func entryPaths(a: [String: Any], b: [String: Any], base: [String: Any]?) -> [[String]] {
        var paths: [[String]] = []

        for key in union(a, b, base, at: "sites").sorted() {
            paths.append(["sites", key])
        }

        for machineId in union(a, b, base, at: "machines").sorted() {
            let inA = element("machines", machineId, a)
            let inB = element("machines", machineId, b)
            let inBase = base.flatMap { element("machines", machineId, $0) }
            // A machine that exists on only one side is one decision, not a dozen.
            guard inA != nil, inB != nil else {
                paths.append(["machines", machineId])
                continue
            }
            // The machine's own scalar fields, minus the collections handled below.
            paths.append(["machines", machineId, "identity"])
            paths.append(["machines", machineId, "wake"])
            for endpoint in union(inA, inB, inBase, at: "endpoints").sorted() {
                paths.append(["machines", machineId, "endpoints", endpoint])
            }
            for system in union(inA, inB, inBase, at: "systems").sorted() {
                paths.append(["machines", machineId, "systems", system])
            }
        }

        paths.append(["appUpdates"])
        return paths
    }

    /// The ids present in an array-of-objects field, across all three documents.
    private static func union(_ a: [String: Any]?, _ b: [String: Any]?, _ base: [String: Any]?,
                              at key: String) -> Set<String> {
        var ids = Set<String>()
        for document in [a, b, base] {
            guard let list = document?[key] as? [[String: Any]] else { continue }
            for entry in list {
                if let id = entry["id"] as? String { ids.insert(id) }
            }
        }
        return ids
    }

    private static func element(_ key: String, _ id: String, _ document: [String: Any]) -> [String: Any]? {
        (document[key] as? [[String: Any]])?.first { $0["id"] as? String == id }
    }

    /// What is at a path, or nil when nothing is.
    static func value(at path: [String], in document: [String: Any]) -> Any? {
        switch path.count {
        case 1:
            return document[path[0]]
        case 2:
            return element(path[0], path[1], document)
        case 3 where path[2] == "identity":
            // Everything about the machine except the collections, which have their own paths.
            guard var machine = element(path[0], path[1], document) else { return nil }
            for key in ["endpoints", "systems", "wake"] { machine.removeValue(forKey: key) }
            return machine
        case 3:
            return element(path[0], path[1], document)?[path[2]]
        case 4:
            guard let machine = element(path[0], path[1], document) else { return nil }
            return (machine[path[2]] as? [[String: Any]])?.first { $0["id"] as? String == path[3] }
        default:
            return nil
        }
    }

    /// The document with one path replaced. Removing is what a nil value means.
    static func setting(_ path: [String], to newValue: Any?, in document: [String: Any]) -> [String: Any] {
        var result = document
        switch path.count {
        case 1:
            if let newValue { result[path[0]] = newValue } else { result.removeValue(forKey: path[0]) }
        case 2:
            var list = (result[path[0]] as? [[String: Any]]) ?? []
            list.removeAll { $0["id"] as? String == path[1] }
            if let entry = newValue as? [String: Any] { list.append(entry) }
            result[path[0]] = list
        case 3 where path[2] == "identity":
            var list = (result[path[0]] as? [[String: Any]]) ?? []
            guard let index = list.firstIndex(where: { $0["id"] as? String == path[1] }) else { return result }
            let keep = list[index]
            var merged = (newValue as? [String: Any]) ?? [:]
            for key in ["endpoints", "systems", "wake"] where keep[key] != nil { merged[key] = keep[key] }
            list[index] = merged
            result[path[0]] = list
        case 3:
            var list = (result[path[0]] as? [[String: Any]]) ?? []
            guard let index = list.firstIndex(where: { $0["id"] as? String == path[1] }) else { return result }
            var machine = list[index]
            if let newValue { machine[path[2]] = newValue } else { machine.removeValue(forKey: path[2]) }
            list[index] = machine
            result[path[0]] = list
        case 4:
            var list = (result[path[0]] as? [[String: Any]]) ?? []
            guard let index = list.firstIndex(where: { $0["id"] as? String == path[1] }) else { return result }
            var machine = list[index]
            var inner = (machine[path[2]] as? [[String: Any]]) ?? []
            inner.removeAll { $0["id"] as? String == path[3] }
            if let entry = newValue as? [String: Any] { inner.append(entry) }
            machine[path[2]] = inner
            list[index] = machine
            result[path[0]] = list
        default:
            break
        }
        return result
    }

    // MARK: - Comparing and printing

    /// Two JSON values, compared by their canonical serialisation. Key order and number formatting
    /// therefore cannot make two identical entries look different.
    static func equal(_ a: Any?, _ b: Any?) -> Bool {
        switch (a, b) {
        case (nil, nil): return true
        case (nil, _), (_, nil): return false
        default: return canonicalString(a!) == canonicalString(b!)
        }
    }

    private static func canonicalString(_ value: Any) -> String {
        let wrapped = ["v": value]
        guard let data = try? JSONSerialization.data(withJSONObject: wrapped, options: [.sortedKeys]) else {
            return String(describing: value)
        }
        return String(decoding: data, as: UTF8.self)
    }

    static func pretty(_ value: Any) -> String {
        let wrapped = ["v": value]
        guard let data = try? JSONSerialization.data(withJSONObject: wrapped,
                                                     options: [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]),
              let text = String(data: data, encoding: .utf8)
        else { return String(describing: value) }
        // Unwrap the one-key object the serialiser needed to accept a bare value.
        let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
        guard lines.count > 2 else { return text }
        return lines.dropFirst().dropLast()
            .map { line -> String in
                var trimmed = String(line)
                if trimmed.hasPrefix("  ") { trimmed.removeFirst(2) }
                return trimmed
            }
            .joined(separator: "\n")
            .replacingOccurrences(of: "\"v\" : ", with: "")
    }

    private static func object(_ data: Data, what: String) throws -> [String: Any] {
        guard let object = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else {
            throw MergeFailure("\(what) is not a JSON object, so the two cannot be compared.")
        }
        return object
    }
}
